### PR TITLE
6407: fix null pointer exception in Page Editor

### LINKED ISSUE #6407: NPE in Page Editor
https://pixiebrix.slack.com/archives/C042T2TBZ17/p1694048778345689

### PR DESCRIPTION
**Pull request by grahamlangford**
## What does this PR do?

- Fixes #6407 

## Reviewer Tips

- New files (`OwnerLabel.tsx`, `useHasEditPermissions.ts`, and `useSortOrganization.ts`) were duplicated code pulled from `PublishContentLayout.tsx` and `ShareRecipeModalBody.tsx`.
- I refactored the files to make it easier to understand how `getScopeAndId` was being used.
- The current fix is in `registryUtils.ts`.

## Checklist

- [ ] Add tests
- [ ] Designate a primary reviewer

### CODE REVIEW

File: src/utils/registryUtils.ts [hunk 1]
```diff
@@ -41,8 +41,14 @@ export function generatePackageId(scope: string, label: string): RegistryId {
  * @param value the full RegistryId
  */
 export function getScopeAndId(
```
Comment (grahamlangford) [inline src/utils/registryUtils.ts:43]: We currently use `getScopeAndId` in three components:
* `ShareRecipeModalBody` (now via the `useHasEditPermissions` hook and `OwnerLabel`
* `PublishContentLayout` (also via the above hook and component)
* `RegistryIdWidget`

Both modals handle `undefined` values already.

`RegistryIdWidget` does not handle the second index being `undefined` well. 
See: https://github.com/pixiebrix/pixiebrix-extension/blob/d009b4a36d87c0113f928b725705a4daa09ed576/src/components/form/widgets/RegistryIdWidget.tsx#L58-L75

`idValue` could already be undefined though, so this change may be fine?

File: src/components/form/widgets/RegistryIdWidget.tsx [hunk 2]
```diff
@@ -51,7 +51,8 @@ const RegistryIdWidget: React.VFC<{
       (organization) =>
         !isEmpty(organization.scope) && editorRoles.has(organization.role)
     )
-    .map((organization) => organization.scope);
+    .map((organization) => organization.scope)
+    .filter(Boolean); // Scope is optional
```
Comment (BLoe) [inline src/components/form/widgets/RegistryIdWidget.tsx:55]: Can also use `compact()` from lodash on the whole result.

Comment (twschiller) [reply src/components/form/widgets/RegistryIdWidget.tsx:55]: Why is this line necessary? There's a `!isEmpty(organization.scope)` check before the map

File: src/extensionConsole/pages/mods/modals/shareModals/OwnerLabel.tsx [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  if (ownerOrganizationIndex === -1) {
+    return (
+      <span>
+        <FontAwesomeIcon icon={faUsers} /> Unknown
+      </span>
+    );
+  }
+
+  // We get the owner's organization and remove it from the list of organizations (splice mutates the array)
+  const ownerOrganization = sortedOrganizations.splice(
```
Comment (twschiller) [inline src/extensionConsole/pages/mods/modals/shareModals/OwnerLabel.tsx:59]: I know this was in the original code, but it seems odd to be mutating the result of the useSortOrganizations hook in the render method

Reading the code, couldn't we just use `find` in this method?

File: src/utils/registryUtils.ts [hunk 4]
```diff
@@ -41,8 +41,14 @@ export function generatePackageId(scope: string, label: string): RegistryId {
...
-  value: RegistryId
+  value?: RegistryId
```
Comment (twschiller) [inline src/utils/registryUtils.ts:44]: This should be `value: Registry | null`. It never makes sense to call `getScopeAndId` with no argument

File: src/extensionConsole/pages/mods/modals/shareModals/useSortOrganizations.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import { selectAuth } from "@/auth/authSelectors";
+import { type Organization } from "@/types/contract";
+import { sortBy } from "lodash";
+import { useMemo } from "react";
+import { useSelector } from "react-redux";
+
+const sortOrganizations = (organizations: Organization[]) =>
```
Comment (twschiller) [inline src/extensionConsole/pages/mods/modals/shareModals/useSortOrganizations.ts:24]: NIT: there's no reason to have this as it's own method defined at module level. Might as well put it inline in the useMemo

File: src/extensionConsole/pages/mods/modals/shareModals/useHasEditPermissions.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+  const sortedOrganizations = useSortOrganizations();
+  const [recipeScope] = getScopeAndId(recipe?.metadata.id);
+
+  let hasEditPermissions = false;
+
+  if (recipeScope === userScope) {
+    hasEditPermissions = true;
+  } else {
+    const ownerOrganizationIndex = sortedOrganizations.findIndex(
```
Comment (twschiller) [inline src/extensionConsole/pages/mods/modals/shareModals/useHasEditPermissions.ts:42]: This is another place where it's not clear to me why we aren't using `find`. Why do we need to remove it from the array?

File: src/extensionConsole/pages/mods/modals/shareModals/useHasEditPermissions.ts [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+      (x) => x.scope === recipeScope
+    );
+
+    if (ownerOrganizationIndex !== -1) {
+      const ownerOrganization = sortedOrganizations.splice(
+        ownerOrganizationIndex,
+        1
+      )[0];
+
+      hasEditPermissions = editorRoles.has(ownerOrganization?.role);
```
Comment (twschiller) [inline src/extensionConsole/pages/mods/modals/shareModals/useHasEditPermissions.ts:52]: At this point, ownerOrganization must be defined due to the ownerOrganizationIndex check above

**Review by BLoe — APPROVED** [review]

File: src/components/form/widgets/RegistryIdWidget.tsx (hunk 2 above)
Comment (BLoe) [reply src/components/form/widgets/RegistryIdWidget.tsx:55]: Oh huh that's true... How did we miss that @grahamlangford? 🤦‍♂️

File: src/extensionConsole/pages/mods/modals/shareModals/OwnerLabel.tsx (hunk 3 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/mods/modals/shareModals/OwnerLabel.tsx:59]: That's a very good point. I'll make the change.

File: src/components/form/widgets/RegistryIdWidget.tsx (hunk 2 above)
Comment (grahamlangford) [reply src/components/form/widgets/RegistryIdWidget.tsx:55]: I don't know. I'm removing it.

File: src/extensionConsole/pages/mods/modals/shareModals/useHasEditPermissions.ts (hunk 6 above)
Comment (grahamlangford) [reply src/extensionConsole/pages/mods/modals/shareModals/useHasEditPermissions.ts:42]: I was focusing merely on eliminating duplicate code to help me find the actual repercussions of the change to `getScopeAndId`. I didn't strongly analyze the other code. I agree that `find` is better here. Making the change.

File: src/extensionConsole/pages/mods/modals/shareModals/OwnerLabel.tsx [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+
+const OwnerLabel: React.FunctionComponent = () => {
+  const { blueprintId } = useSelector(selectShowPublishContext);
+  const { scope: userScope } = useSelector(selectAuth);
+
+  const { data: recipe } = useOptionalModDefinition(blueprintId);
+
+  const sortedOrganizations = useSortOrganizations();
+
+  const [recipeScope] = getScopeAndId(recipe?.metadata.id);
```
Comment (grahamlangford) [inline src/extensionConsole/pages/mods/modals/shareModals/OwnerLabel.tsx:36]: Calling `getScopeAndId` with a nullish `recipe` here results in it being called with `undefined`

File: src/utils/registryUtils.ts (hunk 4 above)
Comment (grahamlangford) [reply src/utils/registryUtils.ts:44]: I can change the typing to `value?: Registry | null` to cover both cases though.

**Pull request merged**